kernel/src/syscall.rs

```diff
@@ -437,6 +408,14 @@ pub enum SyscallReturn {
     /// Generic success case, with an additional 32-bit and 64-bit data field
     SuccessU32U64(u32, u64),
 
+    /// Generic success case, with an additional usize data field
+    SuccessUsize(usize),
```

Comment:
Amit and I chatted some about this out-of-band, but to summarize, I think the realization I've come to from https://github.com/tock/tock/pull/4228 is that `usize` is a Rust-ism, and as such the name USize and its semantics should not be in a language-agnostic syscall ABI. Indeed, [`usize` is problematic here already](https://github.com/tock/tock/pull/4174/files#r1837140155) as well as the [broader questions upstream rust seems to be dealing with around usize](https://github.com/tock/tock/pull/4228#issuecomment-2475014886).

We should just avoid all the headache by using our own, explicit pointer type — which we already have defined here, `SuccessPtr`. In particular, I don't think we should revert stuff like the [removing `as u32` in memop](https://github.com/tock/tock/pull/4174/files#diff-72cf39306bd1015f33c14311acd48af0d0646ee87bb82255e62ef71d12c4e54aL59), but rather those should _also_ be returning `SuccessPtr`s which inside have `CapabilityPtr`s with no capabilities (since that's what they really are, IIUC).

I don't think (I hope not) this is just a bikeshed around the use of the name `usize`. I think it's important to switch all the pointers to pointer types here to make sure that actually works as I think it should and doesn't miss corner cases somehow.

I think Leon's abstracting out the `toTRD104compliant` really helps this, because for all of the existing arch's, `SuccessPtr -> SuccessU32` is a nop cast. For any 64-bit or CHERI platforms, we do not (yet) have a formally defined ABI, so those implementations can do what they like with `SuccessPtr` — I believe/hope that this should not require much, if any, internal churn with the layering we now have in place.

Reply:
So, I agree rusts `usize` may end up as a rust-ism and is therefore not something to expose in a language agnostic way.

But, the type this is meant to represent in a language agnostic way is _not_ a pointer. It is an address. `vaddr_t` is used in some C libraries to seperate this from `size_t`. It has no provenance, and carries no authority on CHERI systems. We should not replace `SuccessUsize` with `SuccessPtr`, even if we bikeshed `SuccessUsize`

In the C ABI, the best type for an address is `size_t`. With what I am hoping rustc will become w.r.t. CHERI, I think it will be `usize`. If rust went another way, and defined another new explit `addr` type, I would at that point expect a kernel refactor to use _that_ for what SucessUSize wraps. Which possibly means it could be named better.

I don't think we really need a "size of object" and "address" type, because given that objects can be as large as the address space but not larger they really are the same type. I am happy to rename `SucessUsize` to make it sound less like the rust type `usize`, or C type `size_t`. But, as it stands, that's what it is.

Reply:
With this being `SuccessAddr` now, can we resolve this thread?